File: src/bar/bar.rs
use super::blocks::Block;
use super::font::{Font, FontDraw};
use crate::Config;
use crate::errors::X11Error;
use std::time::Instant;
use x11rb::COPY_DEPTH_FROM_PARENT;
use x11rb::connection::Connection;
use x11rb::protocol::xproto::*;
use x11rb::rust_connection::RustConnection;

pub struct Bar {
    window: Window,
    width: u16,
    height: u16,
    graphics_context: Gcontext,

    font_draw: FontDraw,

    tag_widths: Vec<u16>,
    needs_redraw: bool,

    blocks: Vec<Box<dyn Block>>,
    block_last_updates: Vec<Instant>,
    block_underlines: Vec<bool>,
    status_text: String,

    tags: Vec<String>,
    scheme_normal: crate::ColorScheme,
    scheme_occupied: crate::ColorScheme,
    scheme_selected: crate::ColorScheme,
}

impl Bar {
    pub fn new(
        connection: &RustConnection,
        screen: &Screen,
        screen_num: usize,
        config: &Config,
        display: *mut x11::xlib::Display,
        font: &Font,
        x: i16,
        y: i16,
        width: u16,
    ) -> Result<Self, X11Error> {
        let window = connection.generate_id()?;
        let graphics_context = connection.generate_id()?;

        let height = (font.height() as f32 * 1.4) as u16;

        connection.create_window(
            COPY_DEPTH_FROM_PARENT,
            window,
            screen.root,
            x,
            y,
            width,
            height,
            0,
            WindowClass::INPUT_OUTPUT,
            screen.root_visual,
            &CreateWindowAux::new()
                .background_pixel(config.scheme_normal.background)
                .event_mask(EventMask::EXPOSURE | EventMask::BUTTON_PRESS)
                .override_redirect(1),
        )?;

        connection.create_gc(
            graphics_context,
            window,
            &CreateGCAux::new()
                .foreground(config.scheme_normal.foreground)
                .background(config.scheme_normal.background),
        )?;

        connection.map_window(window)?;
        connection.flush()?;

        let visual = unsafe { x11::xlib::XDefaultVisual(display, screen_num as i32) };
        let colormap = unsafe { x11::xlib::XDefaultColormap(display, screen_num as i32) };

        let font_draw = FontDraw::new(display, window as x11::xlib::Drawable, visual, colormap)?;

        let horizontal_padding = (font.height() as f32 * 0.4) as u16;

        let tag_widths = config
            .tags
            .iter()
            .map(|tag| {
                let text_width = font.text_width(tag);
                text_width + (horizontal_padding * 2)
            })
            .collect();

        let blocks: Vec<Box<dyn Block>> = config
            .status_blocks
            .iter()
            .map(|block_config| block_config.to_block())
            .collect();

        let block_underlines: Vec<bool> = config
            .status_blocks
            .iter()
            .map(|block_config| block_config.underline)
            .collect();

        let block_last_updates = vec![Instant::now(); blocks.len()];

        Ok(Bar {
            window,
            width,
            height,
            graphics_context,
            font_draw,
            tag_widths,
            needs_redraw: true,
            blocks,
            block_last_updates,
            block_underlines,
            status_text: String::new(),
            tags: config.tags.clone(),
            scheme_normal: config.scheme_normal,
            scheme_occupied: config.scheme_occupied,
            scheme_selected: config.scheme_selected,
        })
    }

    pub fn window(&self) -> Window {
        self.window
    }

    pub fn height(&self) -> u16 {
        self.height
    }

    pub fn invalidate(&mut self) {
        self.needs_redraw = true;
    }

    pub fn update_blocks(&mut self) {
        let now = Instant::now();
        let mut changed = false;

        for (i, block) in self.blocks.iter_mut().enumerate() {
            let elapsed = now.duration_since(self.block_last_updates[i]);

            if elapsed >= block.interval() {
                if block.content().is_ok() {
                    self.block_last_updates[i] = now;
                    changed = true;
                }
            }
        }

        if changed {
            let mut parts = Vec::new();
            for block in &mut self.blocks {
                if let Ok(text) = block.content() {
                    parts.push(text);
                }
            }
            self.status_text = parts.join("");
            self.needs_redraw = true;
        }
    }

    pub fn draw(
        &mut self,
        connection: &RustConnection,
        font: &Font,
        display: *mut x11::xlib::Display,
        current_tags: u32,
        occupied_tags: u32,
        draw_blocks: bool,
        layout_symbol: &str,
        keychord_indicator: Option<&str>,
    ) -> Result<(), X11Error> {
        if !self.needs_redraw {
            return Ok(());
        }

        connection.change_gc(
            self.graphics_context,
            &ChangeGCAux::new().foreground(self.scheme_normal.background),
        )?;
        connection.poly_fill_rectangle(
            self.window,
            self.graphics_context,
            &[Rectangle {
                x: 0,
                y: 0,
                width: self.width,
                height: self.height,
            }],
        )?;

        let mut x_position: i16 = 0;

        for (tag_index, tag) in self.tags.iter().enumerate() {
            let tag_mask = 1 << tag_index;
            let is_selected = (current_tags & tag_mask) != 0;
            let is_occupied = (occupied_tags & tag_mask) != 0;

            let tag_width = self.tag_widths[tag_index];

            let scheme = if is_selected {
                &self.scheme_selected
            } else if is_occupied {
                &self.scheme_occupied
            } else {
                &self.scheme_normal
            };

            let text_width = font.text_width(tag);
            let text_x = x_position + ((tag_width - text_width) / 2) as i16;

            let top_padding = 4;
            let text_y = top_padding + font.ascent();

            self.font_draw
                .draw_text(font, scheme.foreground, text_x, text_y, tag);

            if is_selected {
                let font_height = font.height();
                let underline_height = font_height / 8;
                let bottom_gap = 3;
                let underline_y = self.height as i16 - underline_height as i16 - bottom_gap;

                let underline_padding = 4;
                let underline_width = tag_width - underline_padding;
                let underline_x = x_position + (underline_padding / 2) as i16;

                connection.change_gc(
                    self.graphics_context,
                    &ChangeGCAux::new().foreground(scheme.underline),
                )?;
                connection.poly_fill_rectangle(
                    self.window,
                    self.graphics_context,
                    &[Rectangle {
                        x: underline_x,
                        y: underline_y,
                        width: underline_width,
                        height: underline_height,
                    }],
                )?;
            }

            x_position += tag_width as i16;
        }

        x_position += 10;

        let text_x = x_position;
        let top_padding = 4;
        let text_y = top_padding + font.ascent();

        self.font_draw.draw_text(
            font,
            self.scheme_normal.foreground,
            text_x,
            text_y,
            layout_symbol,
        );

        x_position += font.text_width(layout_symbol) as i16;

        if let Some(indicator) = keychord_indicator {
            x_position += 10;

            let text_x = x_position;
            let text_y = top_padding + font.ascent();

            self.font_draw.draw_text(
                font,
                self.scheme_selected.foreground,
                text_x,
                text_y,
                indicator,
            );
        }

        if draw_blocks && !self.status_text.is_empty() {
            let padding = 10;
            let mut x_position = self.width as i16 - padding;

            for (i, block) in self.blocks.iter_mut().enumerate().rev() {
                if let Ok(text) = block.content() {
                    let text_width = font.text_width(&text);
                    x_position -= text_width as i16;

                    let top_padding = 4;
                    let text_y = top_padding + font.ascent();

                    self.font_draw
                        .draw_text(font, block.color(), x_position, text_y, &text);

                    if self.block_underlines[i] {
                        let font_height = font.height();
                        let underline_height = font_height / 8;
                        let bottom_gap = 3;
                        let underline_y = self.height as i16 - underline_height as i16 - bottom_gap;

                        let underline_padding = 8;
                        let underline_width = text_width + underline_padding;
                        let underline_x = x_position - (underline_padding / 2) as i16;

                        connection.change_gc(
                            self.graphics_context,
                            &ChangeGCAux::new().foreground(block.color()),
                        )?;

                        connection.poly_fill_rectangle(
                            self.window,
                            self.graphics_context,
                            &[Rectangle {
                                x: underline_x,
                                y: underline_y,
                                width: underline_width,
                                height: underline_height,
                            }],
                        )?;
                    }
                }
            }
        }
        connection.flush()?;
        unsafe {
            x11::xlib::XFlush(display);
        }
        self.needs_redraw = false;

        Ok(())
    }

    pub fn handle_click(&self, click_x: i16) -> Option<usize> {
        let mut current_x_position = 0;

        for (tag_index, &tag_width) in self.tag_widths.iter().enumerate() {
            if click_x >= current_x_position && click_x < current_x_position + tag_width as i16 {
                return Some(tag_index);
            }
            current_x_position += tag_width as i16;
        }
        None
    }

    pub fn needs_redraw(&self) -> bool {
        self.needs_redraw
    }
}

File: src/bar/bar.zig
const std = @import("std");
const xlib = @import("../x11/xlib.zig");
const monitor_mod = @import("../monitor.zig");
const client_mod = @import("../client.zig");
const blocks_mod = @import("blocks/blocks.zig");
const config_mod = @import("../config/config.zig");
const ColorScheme = config_mod.ColorScheme;

const Monitor = monitor_mod.Monitor;
const Block = blocks_mod.Block;

fn get_layout_symbol(layout_index: u32) []const u8 {
    const cfg = config_mod.get_config();
    if (cfg) |conf| {
        return switch (layout_index) {
            0 => conf.layout_tile_symbol,
            1 => conf.layout_monocle_symbol,
            2 => conf.layout_floating_symbol,
            3 => conf.layout_scrolling_symbol,
            4 => conf.layout_grid_symbol,
            else => "[?]",
        };
    }
    return switch (layout_index) {
        0 => "[]=",
        1 => "[M]",
        2 => "><>",
        3 => "[S]",
        4 => "[#]",
        else => "[?]",
    };
}

pub const Bar = struct {
    window: xlib.Window,
    pixmap: xlib.Pixmap,
    graphics_context: xlib.GC,
    xft_draw: ?*xlib.XftDraw,
    width: i32,
    height: i32,
    monitor: *Monitor,

    font: ?*xlib.XftFont,
    font_height: i32,

    scheme_normal: ColorScheme,
    scheme_selected: ColorScheme,
    scheme_occupied: ColorScheme,
    scheme_urgent: ColorScheme,
    hide_vacant_tags: bool,

    allocator: std.mem.Allocator,
    blocks: std.ArrayList(Block),
    needs_redraw: bool,
    next: ?*Bar,

    pub fn create(
        allocator: std.mem.Allocator,
        display: *xlib.Display,
        screen: c_int,
        monitor: *Monitor,
        config: config_mod.Config,
    ) ?*Bar {
        const bar = allocator.create(Bar) catch return null;

        const visual = xlib.XDefaultVisual(display, screen);
        const colormap = xlib.XDefaultColormap(display, screen);
        const depth = xlib.XDefaultDepth(display, screen);
        const root = xlib.XRootWindow(display, screen);

        const font_name_z = allocator.dupeZ(u8, config.font) catch return null;
        defer allocator.free(font_name_z);

        const font = xlib.XftFontOpenName(display, screen, font_name_z);
        if (font == null) {
            allocator.destroy(bar);
            return null;
        }

        const font_height = font.*.ascent + font.*.descent;
        const bar_height: i32 = @intCast(@as(i32, font_height) + 8);

        const window = xlib.c.XCreateSimpleWindow(
            display,
            root,
            monitor.mon_x,
            monitor.mon_y,
            @intCast(monitor.mon_w),
            @intCast(bar_height),
            0,
            0,
            0x1a1b26,
        );

        _ = xlib.c.XSetWindowAttributes{};
        var attributes: xlib.c.XSetWindowAttributes = undefined;
        attributes.override_redirect = xlib.True;
        attributes.event_mask = xlib.c.ExposureMask | xlib.c.ButtonPressMask;
        _ = xlib.c.XChangeWindowAttributes(display, window, xlib.c.CWOverrideRedirect | xlib.c.CWEventMask, &attributes);

        const pixmap = xlib.XCreatePixmap(
            display,
            window,
            @intCast(monitor.mon_w),
            @intCast(bar_height),
            @intCast(depth),
        );

        const graphics_context = xlib.XCreateGC(display, pixmap, 0, null);

        const xft_draw = xlib.XftDrawCreate(display, pixmap, visual, colormap);

        _ = xlib.XMapWindow(display, window);

        bar.* = Bar{
            .window = window,
            .pixmap = pixmap,
            .graphics_context = graphics_context,
            .xft_draw = xft_draw,
            .width = monitor.mon_w,
            .height = bar_height,
            .monitor = monitor,
            .font = font,
            .font_height = font_height,
            .scheme_normal = config.scheme_normal,
            .scheme_selected = config.scheme_selected,
            .scheme_occupied = config.scheme_occupied,
            .scheme_urgent = config.scheme_urgent,
            .hide_vacant_tags = config.hide_vacant_tags,
            .allocator = allocator,
            .blocks = .{},
            .needs_redraw = true,
            .next = null,
        };

        monitor.bar_win = window;
        monitor.win_y = monitor.mon_y + bar_height;
        monitor.win_h = monitor.mon_h - bar_height;

        return bar;
    }

    pub fn destroy(self: *Bar, allocator: std.mem.Allocator, display: *xlib.Display) void {
        if (self.xft_draw) |xft_draw| {
            xlib.XftDrawDestroy(xft_draw);
        }
        if (self.font) |font| {
            xlib.XftFontClose(display, font);
        }
        _ = xlib.XFreeGC(display, self.graphics_context);
        _ = xlib.XFreePixmap(display, self.pixmap);
        _ = xlib.c.XDestroyWindow(display, self.window);
        self.blocks.deinit(self.allocator);
        allocator.destroy(self);
    }

    pub fn add_block(self: *Bar, block: Block) void {
        self.blocks.append(self.allocator, block) catch {};
    }

    pub fn invalidate(self: *Bar) void {
        self.needs_redraw = true;
    }

    pub fn draw(self: *Bar, display: *xlib.Display, tags: []const []const u8) void {
        if (!self.needs_redraw) return;

        self.fill_rect(display, 0, 0, self.width, self.height, self.scheme_normal.background);

        var x_position: i32 = 0;
        const padding: i32 = 8;
        const monitor = self.monitor;
        const current_tags = monitor.tagset[monitor.sel_tags];

        for (tags, 0..) |tag, index| {
            const tag_mask: u32 = @as(u32, 1) << @intCast(index);
            const is_selected = (current_tags & tag_mask) != 0;
            const is_occupied = has_clients_on_tag(monitor, tag_mask);

            if (self.hide_vacant_tags and !is_occupied and !is_selected) continue;

            const scheme = if (is_selected) self.scheme_selected else if (is_occupied) self.scheme_occupied else self.scheme_normal;

            const tag_text_width = self.text_width(display, tag);
            const tag_width = tag_text_width + padding * 2;

            if (is_selected) {
                self.fill_rect(display, x_position, self.height - 3, tag_width, 3, scheme.border);
            }

            const text_y = @divTrunc(self.height + self.font_height, 2) - 4;
            self.draw_text(display, x_position + padding, text_y, tag, scheme.foreground);

            x_position += tag_width;
        }

        x_position += padding;

        const layout_symbol = get_layout_symbol(monitor.sel_lt);
        self.draw_text(display, x_position, @divTrunc(self.height + self.font_height, 2) - 4, layout_symbol, self.scheme_normal.foreground);
        x_position += self.text_width(display, layout_symbol) + padding;

        var block_x: i32 = self.width - padding;
        var block_index: usize = self.blocks.items.len;
        while (block_index > 0) {
            block_index -= 1;
            const block = &self.blocks.items[block_index];
            const content = block.get_content();
            const content_width = self.text_width(display, content);
            block_x -= content_width;
            self.draw_text(display, block_x, @divTrunc(self.height + self.font_height, 2) - 4, content, block.color());
            if (block.underline) {
                self.fill_rect(display, block_x, self.height - 2, content_width, 2, block.color());
            }
            block_x -= padding;
        }

        _ = xlib.XCopyArea(display, self.pixmap, self.window, self.graphics_context, 0, 0, @intCast(self.width), @intCast(self.height), 0, 0);
        _ = xlib.XSync(display, xlib.False);

        self.needs_redraw = false;
    }

    fn fill_rect(self: *Bar, display: *xlib.Display, x: i32, y: i32, width: i32, height: i32, color: c_ulong) void {
        _ = xlib.XSetForeground(display, self.graphics_context, color);
        _ = xlib.XFillRectangle(display, self.pixmap, self.graphics_context, x, y, @intCast(width), @intCast(height));
    }

    fn draw_text(self: *Bar, display: *xlib.Display, x: i32, y: i32, text: []const u8, color: c_ulong) void {
        if (self.xft_draw == null or self.font == null) return;

        var xft_color: xlib.XftColor = undefined;
        var render_color: xlib.XRenderColor = undefined;
        render_color.red = @intCast((color >> 16 & 0xff) * 257);
        render_color.green = @intCast((color >> 8 & 0xff) * 257);
        render_color.blue = @intCast((color & 0xff) * 257);
        render_color.alpha = 0xffff;

        const visual = xlib.XDefaultVisual(display, 0);
        const colormap = xlib.XDefaultColormap(display, 0);

        _ = xlib.XftColorAllocValue(display, visual, colormap, &render_color, &xft_color);

        xlib.XftDrawStringUtf8(self.xft_draw, &xft_color, self.font, x, y, text.ptr, @intCast(text.len));

        xlib.XftColorFree(display, visual, colormap, &xft_color);
    }

    fn text_width(self: *Bar, display: *xlib.Display, text: []const u8) i32 {
        if (self.font == null) return 0;

        var extents: xlib.XGlyphInfo = undefined;
        xlib.XftTextExtentsUtf8(display, self.font, text.ptr, @intCast(text.len), &extents);
        return extents.xOff;
    }

    pub fn handle_click(self: *Bar, click_x: i32, tags: []const []const u8) ?usize {
        var x_position: i32 = 0;
        const padding: i32 = 8;
        const display = xlib.c.XOpenDisplay(null) orelse return null;
        defer _ = xlib.XCloseDisplay(display);

        const monitor = self.monitor;
        const current_tags = monitor.tagset[monitor.sel_tags];
        for (tags, 0..) |tag, index| {
            const tag_mask = @as(u32, 1) << @intCast(index);
            const is_selected = (current_tags & tag_mask) != 0;
            const is_occupied = has_clients_on_tag(monitor, tag_mask);

            if (self.hide_vacant_tags and !is_occupied and !is_selected) continue;

            const tag_text_width = self.text_width(display, tag);
            const tag_width = tag_text_width + padding * 2;

            if (click_x >= x_position and click_x < x_position + tag_width) {
                return index;
            }
            x_position += tag_width;
        }
        return null;
    }

    pub fn update_blocks(self: *Bar) void {
        var changed = false;
        for (self.blocks.items) |*block| {
            if (block.update()) {
                changed = true;
            }
        }
        if (changed) {
            self.needs_redraw = true;
        }
    }

    pub fn clear_blocks(self: *Bar) void {
        self.blocks.clearRetainingCapacity();
    }
};

fn has_clients_on_tag(monitor: *Monitor, tag_mask: u32) bool {
    var current = monitor.clients;
    while (current) |client| {
        if ((client.tags & tag_mask) != 0) {
            return true;
        }
        current = client.next;
    }
    return false;
}

pub var bars: ?*Bar = null;

pub fn create_bars(allocator: std.mem.Allocator, display: *xlib.Display, screen: c_int) void {
    var current_monitor = monitor_mod.monitors;
    while (current_monitor) |monitor| {
        const bar = Bar.create(allocator, display, screen, monitor, "monospace:size=10");
        if (bar) |created_bar| {
            bars = created_bar;
        }
        current_monitor = monitor.next;
    }
}

pub fn draw_bars(display: *xlib.Display, tags: []const []const u8) void {
    var current_monitor = monitor_mod.monitors;
    while (current_monitor) |monitor| {
        _ = monitor;
        if (bars) |bar| {
            bar.draw(display, tags);
        }
        current_monitor = if (current_monitor) |m| m.next else null;
    }
}

pub fn invalidate_bars() void {
    var current = bars;
    while (current) |bar| {
        bar.invalidate();
        current = bar.next;
    }
}

pub fn destroy_bars(allocator: std.mem.Allocator, display: *xlib.Display) void {
    var current = bars;
    while (current) |bar| {
        const next = bar.next;
        bar.destroy(allocator, display);
        current = next;
    }
    bars = null;
}

pub fn window_to_bar(win: xlib.Window) ?*Bar {
    var current = bars;
    while (current) |bar| {
        if (bar.window == win) {
            return bar;
        }
        current = bar.next;
    }
    return null;
}
